### DISCUSSION TITLE
ReverseMapping unflattening via MapFrom selector

### DISCUSSION THREAD

**Discussion by Luk164** [LuckyPennySoftware/AutoMapper#4277] [Ideas]
I have recently had some issues where I was not expecting ReverseMap to also create unflattening automatically because of lack of experience with automapper. When I finally found out what was going on I looked for ways to fix it but I was a bit disappointed. There are currently two ways to do this:

- Split mapping into two separate parts by using CreateMap<A, B> and  CreateMap<B, A>. This would of course mean you lose ALL extra functionality of ReverseMap and have to do it manually if required.
- Use ForPath and call Ignore on the options. This will allow you to selectively say which parts you want unflattened, but it is a bit clunky and requires you to have two, sometimes really long accessor calls like `.ForMember(dest => dest.Foo, opt => opt.MapFrom(src => src.Alpha!.Bravo!.Charlie.Delta))` and then` .ForPath(dest => dest.Alpha!.Bravo!.Charlie.Delta, opt => opt.Ignore())`

This is quite verbose and if the path ever changes you will need to fix it in two places.
Instead I would like to propose a third option. Simply adding an optional boolean parameter into MapFrom that if set to false would prevent unflattening. Something like this:
```CSharp
.ForMember(dest => dest.Foo, opt => opt.MapFrom(src => src.Alpha!.Bravo!.Charlie.Delta, false))
```

P.S.: The problem this caused to me was that a value I modified in the source (Delta in this example) was getting overwritten during mapping by copies of the original value from objects with the flattened parameters. This made it impossible to change the Delta variable as long as one or more objects containing the flattening were present.

**Comment by jbogard** [comment]
ReverseMap is designed for simple scenarios. Might I suggest not using AutoMapper if the behavior is unexpected and difficult to change? Mapping into more complex objects will be inherently more complex so I would avoid it.

**Reply by Luk164** [reply]
I must say I disagree. Yes reversemap is not meant for very complex scenarios, but what I am suggesting is just a simple upgrade that would make the code a little less verbose and easier to maintain. If it was implemented as an optional argument then it wouldn't even be a breaking change.

In my specific case I was flattening one value from a little bit deeper in the object structure so that I did not need to retrieve it from database by ID.

**Reply by jbogard** [reply]
I think if you actually tried to implement this simple upgrade, you'd see how not simple it would be. Just go for the verbose option, it's far, far more obvious to whoever has to maintain the configuration later.
